src/View/ViewLayerData.php

```diff
@@ -0,0 +1,254 @@
+     */
+    public function getIteratorCount(): int
+    {
+        $count = $this->getRawDataValue('count');
+        if (is_numeric($count)) {
+            return $count;
+        }
+        if (is_countable($this->data)) {
+            return count($this->data);
+        }
+        if (ClassInfo::hasMethod($this->data, 'getIterator')) {
+            return iterator_count($this->data->getIterator());
+        }
+        return 0;
+    }
+
+    public function getIterator(): Traversable
+    {
+        if (!is_iterable($this->data) && !ClassInfo::hasMethod($this->data, 'getIterator')) {
+            $type = get_class($this->data);
+            throw new BadMethodCallException("$type is not iterable.");
+        }
+
+        $iterator = $this->data;
+        if (!is_iterable($iterator)) {
+            $iterator = $this->data->getIterator();
+        }
+        $source = $this->data instanceof ModelData ? $this->data : null;
+        foreach ($iterator as $item) {
+            yield $item === null ? null : ViewLayerData::create($item, $source);
+        }
+    }
+
+    /**
+     * Checks if a field is set, or if a getter or a method of that name exists.
+     * We need to check each of these, because we don't currently distinguish between a property, a getter, and a method
+     * which means if any of those exists we have to say the field is "set", otherwise template engines may skip fetching the data.
+     */
+    public function __isset(string $name): bool
+    {
+        // Note we explicitly DO NOT call count() or exists() on the data here because that would
+        // require fetching the data prematurely which could cause performance issues in extreme cases
+        return in_array($name, ViewLayerData::META_DATA_NAMES)
+            || isset($this->data->$name)
+            || ClassInfo::hasMethod($this->data, "get$name")
+            || ClassInfo::hasMethod($this->data, $name);
+    }
+
+    public function __get(string $name): ?ViewLayerData
+    {
+        $value = $this->getRawDataValue($name, type: ViewLayerData::TYPE_PROPERTY);
+        if ($value === null) {
+            return null;
+        }
+        $source = $this->data instanceof ModelData ? $this->data : null;
+        return ViewLayerData::create($value, $source, $name);
+    }
+
+    public function __call(string $name, array $arguments = []): ?ViewLayerData
+    {
+        $value = $this->getRawDataValue($name, $arguments, ViewLayerData::TYPE_METHOD);
+        if ($value === null) {
+            return null;
+        }
+        $source = $this->data instanceof ModelData ? $this->data : null;
+        return ViewLayerData::create($value, $source, $name);
+    }
+
+    public function __toString(): string
+    {
+        if (ClassInfo::hasMethod($this->data, 'forTemplate')) {
+            return $this->data->forTemplate();
+        }
+        return (string) $this->data;
+    }
+
+    /**
+     * Check if there is a truthy value or (for ModelData) if the data exists().
```

Comment:
Can you please say what about it needs rewording so I know what to reword it to? What is unclear about it?

Reply:
Seems like there's a lot of conditionals in the method based on whether you've past in a $name arg, and what $this->data is

`Check if there is a truthy value` is essentially the fallback? Seems like a weird thing to lead the docblock with?

Is this method basically 'Calls $this->data $this->data->exists(), or count($this->data), or casts $this->data to a bool`?  And then it does something else if $name is not null?

Method is confusing. Actually it should probably be split into two different methods, one for where $name is a required param, and another with no params.

Reply:
> Seems like there's a lot of conditionals in the method based on whether you've past in a $name arg

Added a note about the $name condition.

> , and what $this->data is

That's either `ModelData` which is already covered, or an implementation detail that doesn't need to be part of the phpdoc in the case of countable.

> Check if there is a truthy value is essentially the fallback?

It's what is done for non-ModelData. Getting the value is done in a variety of ways (e.g. getting the the count for countable, where 0 is falsy and anything else is truthy) but that's an implementation detail.

> Is this method basically 'Calls $this->data $this->data->exists(), or count($this->data), or casts $this->data to a bool`? And then it does something else if $name is not null?

This method is "calls `exists()` on `ModelData`, and otherwise gets a value in whatever way the implementation dictates and returns true if that value is truthy" Or in other words, "Check if the data has a value" aka "hasDataValue()"